### ISSUE TITLE
[POC] Framework for auto-encoder based approaches for recommendations for NPM

### ISSUE THREAD

**Issue by sara-02** [openshiftio/openshift.io#2004]
## User Story:
As an consumer of Fabric8-Analytics stack report, I should be able to get companion/outlier insights for my NPM stack via the new approach.

## Acceptance Criteria:
Prepare the framework for auto-encoder based approaches as described in [this](https://github.com/openshiftio/openshift.io/issues/2002) issue.
 
## Task List:
- [x] Data Cleaning
- [x] Data Preparation: Generate the Rating Matrix and the Context Matrix
- [x] Feature engineering and selection
- [ ] ~~Create the Supervised Auto-encoder Framework using Keras~~(not required since the CVAE approach worked).
- [ ] ~~Hyper-parameter  tuning~~
- [x] Accuracy evaluation

Note: The details about the approach can be found in [this issue](https://github.com/openshiftio/openshift.io/issues/1802), and in this [doc](https://docs.google.com/document/d/1LTAJRq60lNs-fDGsZLkLS0r8E8j5zDxK_7zkXJqLRxM/edit)

**Comment by rootAvish** [comment]
The CVAE approach worked, and has its results documented here: https://docs.google.com/spreadsheets/d/1OKq1BvPHUzKwYrL8XFWZAgkKC7jcaW0QyY0vleq0wV0/edit?usp=sharing, we did not go forward with implementing the supervised autoencoder approach as it would perform inferior to this approach and was a fallback in case the other did not work. Work around creating a post-filter business logic for the same  will be carried out in this sprint.
--
